README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Bootstrapping and kinship estimation

## Installation

To get the latest version, install from GitHub as follows:

``` r
 # First install devtools if needed
if(!require(devtools)) install.packages("devtools")

# Install dvir from GitHub
devtools::install_github("thoree/kinBoot")
```

Currently the main function is `bootPhi`. The purpose of the function is
to compare parametric and nonparametric boostrapping for the kinship
coefficient. It is a wrapper for some functions of the `ped suite` of
R-libraries, in particular **forrel**. The below figure (to be repaced
by a proper figure) explains the simulation experiement.

![Simulation experiment explained](figures/simExplained.png)

## Tutorial example

The goal of this example is to explain in detail how the simulation
experiment to evaluate and compare parametric and nonparametric
bootstrap is performed. We consider double first cousins. The IBD
coefficient kappa and kinship coefficient phi are calculated below

``` r
library(kinBoot)
#> Loading required package: pedtools
library(forrel)
library(ribd)
# library(coxed) # Only needed for coxed::bca confidence intervals
ped = doubleFirstCousins() 
ids = leaves(ped)
phi = kinship(ped, ids)
c(kappa = kappaIBD(ped, ids), phi = phi)
#> kappa1 kappa2 kappa3    phi 
#> 0.5625 0.3750 0.0625 0.1250
```

We set the 35 markers in the database `forrel::NorwegianFrequencies` as
follows:

``` r
ped = setMarkers(ped, locusAttributes =  NorwegianFrequencies)
```

Below N = 2 simulations are done using only B = 100 bootstraps.

``` r
res1 = bootPhi(ped, ids, N = 2, B = 100, seed = 17)
```

Excerpts of the output is discussed next starting with the parametric
simulations:

``` r
round(res1$simParametric[1:2,], 4)
#>      realised   boot    bias  lower  upper cover
#> sim1   0.2023 0.1978 -0.0046 0.1269 0.2605     0
#> sim2   0.0981 0.0983  0.0002 0.0189 0.1663     1
```

For the first simulation, `sim1` above, marker data were generated for
the double cousins using the function `forrel::profileSim`. Kappa was
estimated by `forrel::IBDestimate` and the realised value 0.2023 found.
The average phi based on the B=100 bootstraps from
`forrel::ibdBootstrap` is 0.1978 and the average bias is -0.0046. The
95% confidence interval was estimated with the default percentile method
using the `quantile`function. The value of `cover` is 0 for `sim1` since
this interval does not cover the kinship coefficient 0.125. The output
for the nonparametric method is obviusly conceptually similar:

``` r
round(res1$simNonparametric[1:2,], 4)
#>      realised   boot    bias  lower  upper cover
#> sim1   0.2023 0.1985 -0.0038 0.1223 0.2687     1
#> sim2   0.0981 0.0977 -0.0004 0.0268 0.1724     1
```

Note that the `realised`value is the same as above. The values averaged
over all, i.e., 2 simulations in this case, are also reported as shown
below

``` r
round(res1$averaged[1:2,], 4)
#>               realised   boot    bias  lower  upper cover
#> parametric      0.1502 0.1480 -0.0022 0.0729 0.2134   0.5
#> nonparametric   0.1502 0.1481 -0.0021 0.0746 0.2206   1.0
```

Below only one simulation is done and IBD triangle plots are shown with
the parametric to the left:

``` r
res1 = bootPhi(ped, ids, N = 1, B = 100, seed = 17, plot = TRUE)
```

![](man/figures/README-unnamed-chunk-9-1.png)<!-- -->

The results from `ibdBootstrap` are returned from the last simulation.
This can be used for further plotting, for instance normality plots as
next

``` r
x = res1$bootParametric
par = 0.25*x$k1 + 0.5*x$k2
x = res1$bootNonparametric
non = 0.25*x$k1 + 0.5*x$k2
par(mfrow = c(1,2))
qqnorm(par, main = "Parametric", xlab ="", ylab ="")
qqnorm(non, main = "Nonparametric", xlab ="", ylab ="")
```

![](man/figures/README-unnamed-chunk-10-1.png)<!-- -->

We can use boxplots or density plots to compare the distribution of the
estimates of the kinship coefficients as below:

``` r
par(mfrow = c(1,2))
boxplot(par, non, names = c("parametric", "nonparam"),
         main = "",
         sub = "Red stapled line: theoretical value")
abline(h = res1$phi, col = 'red', lty = 2)
plot(density(par), sub = "",  lty = 1, col ="blue", 
     xlab = "kinship, parametric blue", main = "")
points(density(non),  lty = 1, cex = 0.1, col = "red")
```

![](man/figures/README-unnamed-chunk-11-1.png)<!-- -->


NEWS.md
# kinBoot 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package.

* First version 2021-03-26


R/interval.R
#'
#' Bootstrap confidence intervals
#'
#' Some alternative confidence intervals are implemented.
#'
#' @param x Double vector. Bootstrap sample.
#' @param method Character. See details.
#' @param conf.level Double i n(0,1). Confidence level.

#'
#' @return A vector of length 2
#' in which the first element is the lower bound and the second element is the upper bound.
#'
#' @details The implemented methods are `perc`, the standard using the `quantile` function R
#' to find the percentile; `bca`, see `coxed::bca`
#' (his method fails if the input is a vector of constant values), and 'student' based on
#' the normal approximation.
#'
#' @export
#'
#' @examples
#' library(coxed)
#' x = rnorm(100000)
#' interval(x, method = "perc")
#' interval(x, method = "bca")
#' interval(x, method = "student")
#' interval(x, method = "Not implemented")
#'
#' # Example Efron & Tibshirani p 154
#' x = c(52, 104, 146, 10, 51, 30, 40, 27, 46)
#' interval(x, method = "bca", conf.level = 0.9)


interval = function(x, method = "perc", conf.level = 0.95){
  low = (1 - conf.level)/2
  high = 1 - low
  if(method == "perc")
    interval = quantile(x, c(low,  high))
  else if(method == "bca")
    interval = coxed::bca(x, conf.level = conf.level)
  else if(method == "student"){
      est = mean(x)
      n = length(x)
      s = sd(x)
      tq = qt(low,df = n-1,lower.tail = FALSE)
      interval = c(est - tq*s, est + tq*s)
    }
  else  stop(method)
  interval
}







R/examplePhi.R
#'
#' Simulation experiment
#'
#' We consider parametric and nonparametric bootstrap and the kinship coefficient for some pedigrees.
#'
#' @param peds list of ped objects with allele frequencies.
#' @param idlist list of ids of of pair.
#' @param N Integer. No of simulations.
#' @param B Integer. No of bootstraps.
#' @param CItype Logical
#' @param conf.level Double
#' @param seed Integer
#' @param verbose Logical
#' @param db database
#' @param case Logical. Describes experiment.
#' #'
#' @return Returns a data frame summarising the simulation.
#' @details See [kinBoot::bootPhi()]
#'
#' @export
#'
#' @examples
#' library(forrel)
#' library(ribd)
#' library(coxed) # for bca confidence intervals
#'
#' peds = list(quadHalfFirstCousins(), doubleFirstCousins(), nuclearPed(2),
#'             halfSibPed(), cousinPed(1))
#' names(peds) = c("QHFC",  "DFC", "S", "H", "FC")
#' idlist = lapply(peds, leaves)
#' # phi = unlist(lapply(peds, function(x) kinship(x, leaves(x))))

#' N = 1; B = 4 # Increase to N = 100 and B = 1000, at least
#' seed = 1729

#' # Example 1a Many SNP-s. Intended to meet assumptions well
#' n = 1000 # no of markers
#' p = rep(0.5, n)
#' freq = list()
#' for (i in 1:n)
#' freq[[i]] =  list(afreq = c("1" = p[i], "2" = 1- p[i]))
#' db = freq
#' res1 = examplePhi(peds, idlist, N = N, B = B, seed = seed, db = db, case = "1000snps")
#'
#' # Example 1b NorwegianFrequency. Medium number of markers
#' db = NorwegianFrequencies
#' res2 = examplePhi(peds, idlist, N = N, B = B, seed = seed, db = db, case ="Norwegian")
#'
#' # Example 1c Few markers, 8 CODIS markers
#' codis8 = c("CSF1PO", "D3S1358", "D5S818", "D7S820", "D8S1179", "D13S317", "D16S539", "D18S51")
#' db = NorwegianFrequencies[codis8]
#' res3 = examplePhi(peds, idlist, N = N, B = B, seed = seed, db = db, case = "codis8")

#' # Example 1d
#' db1 = NorwegianFrequencies[c("SE33")]
#' res4 = examplePhi(peds, idlist, N = N, B = B, seed = seed,
#'                   db = db1, case = "db1")
#'
#' # Example 1e
#' n = 10 # no of markers
#' p = 0.5
#' freq = list()
#' for (i in 1:(n-1))
#'      freq[[i]] =  list(afreq = c("1" = p, "2" = 1- p))
#' p = rep(1/100,100)
#' names(p) = 1:100
#' freq[[10]] = list(afreq = p)
#' res5 = examplePhi(peds, idlist, N = N, B = B, seed = seed,
#'                 db = freq, case = "nonid")
#'
#'  res = rbind(res1, res2, res3, res4, res5)
#'  save(res, file = "res27mar.Rdata")



examplePhi <- function(peds, idlist, N = 2, B = 10,  CItype = "perc",
                       conf.level = 0.95, seed = NULL, verbose  = TRUE, db = NULL,
                       case = "foo"){
  if(verbose) cat(date())
  if(!is.null(seed))
    set.seed(seed)
  n = length(peds)
  res = matrix(ncol = 6, nrow = 2*n)
  res = res0 = data.frame()
  for(i in 1:n){
    if(verbose) cat("\n", i, " ")
    peds[[i]] = setMarkers(peds[[i]], locusAttributes = db)
    phi = ribd::kinship(peds[[i]], idlist[[i]])
    res0 =  bootPhi(ped = peds[[i]], ids = idlist[[i]], N = N, B = B,
            CItype = CItype, conf.level = conf.level)$average
    res0 = cbind(phi = rep(phi,2), res0)
    res = rbind(res, res0)
  }

  res = data.frame(case = case, par = rep(c("Y","N"), n), ped = rep(names(peds), each =2), res)
  rownames(res) = NULL
  if(verbose) cat(date(), "\n")
  res
}







R/bootPhi.R
#'
#' Properties of bootstrap confidence intervals for the kinship coefficient
#'
#' The purpose is to compare parametric and nonparametric confidence
#' intervals in kinship applications, currently only for the kinship coefficient.
#'
#' @param ped ped object with allele frequencies.
#' @param ids Id of pair.
#' @param N Integer. No of simulations.
#' @param B Integer. No of bootstraps.
#' @param CItype Logical. See [kinBoot::interval()]
#' @param conf.level Double
#' @param plot Logical
#' @param seed Integer.
#' @param verbose Logical.
#'
#'
#' @return A list with the following sif elements:
#'
#'   * `phi`: The kinship coefficient of the pedigree.
#'
#'   * `averaged` : A data frame with two lines, one for `parametric`
#'   and one for `nonParametric`. The values are averaged over the `N` simulations.
#'   The columns `realised`, `boot`, `bias`, `lower`,  `upper`, and  `cover` are explained
#'   in `Details`.
#'
#'   * `simParametric`. The entries are as for `averaged`
#'   but for each parametric simulation.
#'
#'   * `simNonparametric` The entries are as for `averaged`
#'   but for each nonparametric simulation.
#'
#'   * `bootParametric` The result of the last parametric bootstrap simulation.
#'
#'   * `bootParametric` The result of the last nonparametric bootstrap simulation.

#' @details Marker data are simulated `N` times giving `N` estimates of (kappa0, kappa1, kappa2).
#' For each simulation, the `realised` phi is found. Parametric and nonparametric bootstrapping
#' is done giving `boot` (the averaged kinship coefficient from `B` bootstrap simulations)
#' and `bias` = `realised`- `boot` is found. There are various
#' ways to calculate the confidence interval. One option is `bca` as implemented in `coxed::bca`.
#' This method fails if the input is a vector of constant values. The default is 'perc', the standard percentile interval. The
#' variable `cover` is 1 if the confidence interval contains `phi` and 0 otherwise.
#'
#' @export
#'
#' @examples

#'
#' library(forrel)
#' library(ribd)
#' library(coxed) # for bca confidence intervals
#'
#' # Example 0
#' ped = doubleFirstCousins()
#' ped = setMarkers(ped, locusAttributes =  NorwegianFrequencies)
#' ids = leaves(ped)
#' res1 = bootPhi(ped, ids, N = 1, B = 2)
#'
#' Example 2
#' # The example considers the kinship coefficients between
#' # brothers named `B1` and `B2` using the 35 markers in
#' # forrel::NorwegianFrequencies.
#'
#' ids = c("B1", "B2")
#' ped = nuclearPed(2, children = ids)
#' ped = setMarkers(ped, locusAttributes = NorwegianFrequencies)
#' N = 10 # no of confidence intervals. Increase
#' B = 100 # no of  bootstraps. Increase
#' res1 = bootPhi(ped, ids, N = N, B = B, seed = 17)
#'
#' # Basic output
#' res1[1:2]
#'
#' # Compare parametric and nonparametric estimates
#' y1 = res1$simParametric$boot
#' y2 = res1$simNonparametric$boot
#' boxplot(y1, y2, names = c("parametric", "nonparametric"),
#'         main = "Bootstrap estimates of kinship coefficient",
#'         sub = "Red stapled line: theoretical value")
#' abline(h = res1$phi, col = 'red', lty = 2)

bootPhi <- function(ped, ids = NULL, N, B,  CItype = "perc", conf.level = 0.95,
                    plot = F, seed = NULL,  verbose  = F){
  if(!is.null(seed))
    set.seed(seed)
  tabParam  = tabNonparam = matrix(ncol = 6, nrow = N)
  dn = list(c(paste0("sim", 1:N)),
            c("realised", "boot", "bias", "lower", "upper", "cover"))
  dimnames(tabParam) = dimnames(tabNonparam) = dn
  boot = bias =  rep(NA, N)
  # Simulates N profiles
  sim = profileSim(ped, N = N, ids = ids)
  # Estimates kappa
  kappas.obs = lapply(sim, function(x)
               as.double(ibdEstimate(x, verbose = FALSE)[1,4:6]))
  realised = unlist(lapply(kappas.obs, function(x) 0.25*x[2] + 0.5*x[3]))
  phi = kinship(ped, ids = ids)
  if(plot) par(mfrow = c(1,2))
  for (j in 1:N){
    if(verbose) cat(j, " ")
    bootPar = ibdBootstrap(ped, kappa = kappas.obs[[j]],  N = B, plot = plot)
    phis = 0.25*bootPar[,2] + 0.5*bootPar[,3]
    boot = mean(phis)
    bias = boot - realised[j]
    CI = interval(phis, method = CItype, conf.level = conf.level)
    coverage = (CI[1] <= phi) & (CI[2] >= phi)
    tabParam[j, 1:6] = c(realised[j], boot, bias,  CI, coverage)

    bootNon = ibdBootstrap(sim[[j]], ids, param = "kappa", N = B,
                          method = "nonparametric", plot = plot)
    phis = 0.25*bootNon[,2] + 0.5*bootNon[,3]
    boot = mean(phis)
    bias = boot - realised[j]
    CI = interval(phis, method = CItype, conf.level = conf.level)
    coverage = (CI[1] <= phi) & (CI[2] >= phi)
    tabNonparam[j, 1:6] = c(realised[j], boot, bias,  CI, coverage)
   }

  average = rbind(apply(tabParam, 2, mean),
                  apply(tabNonparam, 2, mean))
  average = data.frame(average)
  rownames(average) = c("parametric", "nonparametric")
  list(phi = phi,
       averaged = as.data.frame(average),
       simParametric = as.data.frame(tabParam),
       simNonparametric = as.data.frame(tabNonparam),
       bootParametric = bootPar,
       bootNonparametric = bootNon)
}





